### ISSUE TITLE
support HDR image

### ISSUE THREAD

**Issue by ifbowen** [SDWebImage/SDWebImage#3553]
wwdc2023, apple support HDR image，When sd can support
https://developer.apple.com/videos/play/wwdc2023/10181/

<img width="886" alt="截屏2023-06-14 14 44 10" src="https://github.com/SDWebImage/SDWebImage/assets/17174309/0def8aa7-0667-4426-b9e8-40b97f741ccb">

**Comment by dreampiggy** [comment]
Tip: SDWebImage use ImageIO for default image decoding handling. And no hardcode about the image format ImageIO supported.

**Comment by dreampiggy** [comment]
I don't think for `Web` image, there are `Gain Map HDR` image avaiable online. If you do have, provide the URL for sample.

At least, only 10bit image (HEIF/AVIF/WebP) can store this information in image format. JPEG is no luck.

This HDR image mostly for the iPhone Photo Library or other Professional use case, which seems not what SDWebImage's main propose

**Comment by dreampiggy** [comment]
Maybe we can support HEIF to not strip the HDR content. But actually if you use the default configuration `SDImageCoderHelper.defaultDecodeSolution = .uikit`, we may use ImageIO && UIImage.decodeForDisplay, which may not lose the HDR content.

**Comment by dreampiggy** [comment]
OK...Maybe for ImageIO we should use the new option https://developer.apple.com/documentation/imageio/kcgimagesourcedecodetohdr/

![img_v2_579b6f66-fd23-471e-ae13-dc5761307e9g](https://github.com/SDWebImage/SDWebImage/assets/6919743/d9326fe7-71ae-49db-a74c-446562d9b5dd)

**Comment by mrousavy** [comment]
Dumb question, but is this for displaying HDR images in brighter colors?

When I open an HDR image in the iOS Photos app, my screen gets really bright. Opening the same image with SDWebImage doesn't increase brightness.

**Comment by dreampiggy** [comment]
> Dumb question, but is this for displaying HDR images in brighter colors?
> 
> When I open an HDR image in the iOS Photos app, my screen gets really bright. Opening the same image with SDWebImage doesn't increase brightness.

Most images on the Web, does not use HEIF
Most HEIF images, does not contains HDR information, mostlly are gain-map HDR (which means, not Standard, it's Apple's own solution)

My goal is to firstly support to control (by you) whether to enable the HDR. You can use the option enum like Apple's https://developer.apple.com/documentation/uikit/uiimagedynamicrange?language=objc 

```
[UIImageDynamicRangeConstrainedHigh](https://developer.apple.com/documentation/uikit/uiimagedynamicrange/uiimagedynamicrangeconstrainedhigh?language=objc)
[UIImageDynamicRangeHigh](https://developer.apple.com/documentation/uikit/uiimagedynamicrange/uiimagedynamicrangehigh?language=objc)
[UIImageDynamicRangeStandard](https://developer.apple.com/documentation/uikit/uiimagedynamicrange/uiimagedynamicrangestandard?language=objc)
[UIImageDynamicRangeUnspecified](https://developer.apple.com/documentation/uikit/uiimagedynamicrange/uiimagedynamicrangeunspecified?language=objc)
```

**Comment by dreampiggy** [comment]
Which means, during decoding (Because the Web's image decoding is controlled by us), you can provide whether to strip down, or keep the HDR context

If you prefers to keep the HDR context, and the image is HDR image, and your image view preferrs to display HDR, all these 3 conditions is true.

You will see the more brightness on the display screen.

**Comment by dreampiggy** [comment]
Note:

Keeping the HDR information has at least 2x RAM usage. 

Because non-HDR image can use RGBA888

HDR image must use RGBA16161616, because the brightness are beyond 256 raw value, need 16-bit to storage

**Comment by mrousavy** [comment]
Interesting stuff, thanks for the insights!

I've only worked with Video HDR before, where SDR is YUV 4:2:0 8bit, and HDR is YUV 4:2:0 10bit.

Love your work man, thanks!

**Comment by dreampiggy** [comment]
For the public API in CoreGraphic Image (CGImage), you can not actually use YUV 4:2:0 image

There are private API called IOSurface based CGImage, which can use raw YUV 4:2:0 to getter lower memory usage. This is the built-in pipeline for Apple's JPEG and HEIF hardware codec decoding.

All other format should always use RGB

So, for those who use RGB, the RGB8 for the SDR image
The RGB16 for HDR image

**Comment by dreampiggy** [comment]
For HDR without alpha channel, we can still use that RGB101010 to fits-in 4 bytes per pixel

compared to exists SDR RGB888, only 25% RAM increment, but you can see more bright and vivid colors

**Comment by jiasongs** [comment]
How can I do it currently, can support HDR by SDWebImage.

For example:
<img width="1055" alt="图片" src="https://github.com/user-attachments/assets/c597543f-d036-4326-93d8-dfecff570b7b">

**Comment by dreampiggy** [comment]
No. There are still no support in ImageIO decoder.

But HDR is supported by SDWebImageHEIFCoder, the third-party heif codec

**Comment by jiasongs** [comment]
> 不。ImageIO 解码器仍然不支持。当您使用 HTTP URL 加载基于增益图的 HDR heic 时，它将仅显示为 SDR 颜色（RGBA8888，而不是 RGB10）
> 
> 但是第三方 heif/avif 编解码器 SDWebImageHEIFCoder 和 SDWebImageAVIFCoder 支持 HDR，它们都支持解码 10 位颜色，并且需要更多 RAM 来渲染

I use SDWebImageHEIFCoder, it does not parse as HDR, Demo:
[Test.zip](https://github.com/user-attachments/files/18005062/Test.zip)
<img width="1002" alt="图片" src="https://github.com/user-attachments/assets/9ff144ec-38ec-4732-8f20-e9f3b4a2920a">

**Comment by dreampiggy** [comment]
> I use SDWebImageHEIFCoder, it does not parse as HDR, Demo:

https://github.com/SDWebImage/SDWebImageHEIFCoder/blob/master/SDWebImageHEIFCoder/Classes/SDImageHEIFCoder.m#L239

I guess this should support HDR 10 bit ?

OK, my fault, it hardcode to use 8 bit here:

https://github.com/SDWebImage/SDWebImageHEIFCoder/blob/master/SDWebImageHEIFCoder/Classes/SDImageHEIFCoder.m#L276

**Comment by dreampiggy** [comment]
For ImageIO, if Apple added the easy to use API, we can add support here.

But since HDR need double RAM for each HDR image, I don't think by default we should allows HDR HEIC. Because `SDWebImage`, is for Web Image mostly. Web world is not full of HDRs in 2024

We need a options (context) like `SDWebImageContextDecodeHDRImage: true` to customize this behavior.

**Comment by jiasongs** [comment]
> 对于ImageIO，如果Apple添加了易于使用的API，我们也可以添加支持。对于其他编码器，如 AVIF（使用 aom/dav1d），我们需要一些更新的编解码器，可以由我实现。
> 
> 但由于 HDR 需要每个 HDR 图像最多双倍的 RAM，因此我认为默认情况下我们不应该解码 HDR。因为 `SDWebImage`，主要用于网页图像。 2024 年网络世界还没有充满 HDR，大多数图像仍然是 JPEG/PNG/WebP。尚未准备好 HDR。
> 
> 网络上唯一的 HDR 图像是 `.heic` 和 `.avif`
> 
> 我们需要一个选项（上下文），例如 `SDWebImageContextDecodeHDRImage: true` 自定义此行为。

Yes, users can set `SDWebImageContextDecodeHDRImage: true` according to certain scenarios, such as when viewing a large image or knowing that it is HDR.

In addition, iOS 18 or Android 15, JPEG can also support HDR.

**Comment by dreampiggy** [comment]
> In addition, iOS 18 or Android 15, JPEG can also support HDR.

It's JPEG-XL, not traditional JPEG. I've also provided a custom coder for this (though I hardcode with SDR on that): https://github.com/SDWebImage/SDWebImageJPEGXLCoder

**Comment by dreampiggy** [comment]
> > In addition, iOS 18 or Android 15, JPEG can also support HDR.
> 
> It's JPEG-XL, not traditional JPEG. I've also provided a custom coder for this (though I hardcode with SDR on that): https://github.com/SDWebImage/SDWebImageJPEGXLCoder

Seems it's HDR ready: https://github.com/SDWebImage/SDWebImageJPEGXLCoder/blob/master/SDWebImageJPEGXLCoder/Classes/SDImageJPEGXLCoder.m#L261

But it's not controllable by user. User can choose to fallback to SDR even if the image data buffer originally encoded with HDR to reduce RAM usage, and compatible because their iPhone screen hardware is not compatible with HDR.

**Comment by dreampiggy** [comment]
Can you make a PR for this ? It's not hard, check my PR for how to  add new coder options. Like this one: https://github.com/SDWebImage/SDWebImage/pull/3537

Because recently I have no much time on open-source project maintainess , review PR and code takes less time

**Comment by jiasongs** [comment]
> 你能为此做一个 PR 吗？这并不难，请查看我的 PR 了解如何添加新的编码器选项。像这个： [ #3537 ](https://github.com/SDWebImage/SDWebImage/pull/3537)
> 
> 因为最近开源项目维护的时间不多，审阅PR和代码的时间比较少

OK, our project has put HDR on hold for now, if I have time, I'll submit a PR

**Comment by jiasongs** [comment]
> 你能为此做一个 PR 吗？这并不难，请查看我的 PR 了解如何添加新的编码器选项。像这个： [ #3537 ](https://github.com/SDWebImage/SDWebImage/pull/3537)
> 
> 因为最近开源项目维护的时间不多，审阅PR和代码的时间比较少

@dreampiggy I have submitted a PR https://github.com/SDWebImage/SDWebImage/pull/3778, no issues have been found in our project yet.
